File: golang/2434_robot_with_string/robot_with_string.go
// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func robotWithString(s string) string {
	cnt := make([]int, 26)
	for _, c := range s {
		cnt[c-'a']++
	}

	stack := []rune{}
	res := []rune{}
	minCharacter := 'a'
	for _, c := range s {
		stack = append(stack, c)
		cnt[c-'a']--
		for minCharacter != 'z' && cnt[minCharacter-'a'] == 0 {
			minCharacter++
		}
		for len(stack) > 0 && stack[len(stack)-1] <= minCharacter {
			res = append(res, stack[len(stack)-1])
			stack = stack[:len(stack)-1]
		}
	}

	return string(res)
}

func main() {
	tests := []struct {
		s   string
		ans string
	}{
		{"zza", "azz"},
		{"bac", "abc"},
		{"bdda", "addb"},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, robotWithString(test.s), index)
	}
}

File: cpp/2434_robot_with_string/robot_with_string.cpp
#include <cassert>
#include <stack>
#include <string>
#include <tuple>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
  string robotWithString(string s) {
    unordered_map<char, int> cnt;
    for (char c : s) {
      cnt[c]++;
    }

    stack<char> stk;
    string res;
    char minCharacter = 'a';
    for (char c : s) {
      stk.emplace(c);
      cnt[c]--;
      while (minCharacter != 'z' && cnt[minCharacter] == 0) {
        minCharacter++;
      }
      while (!stk.empty() && stk.top() <= minCharacter) {
        res.push_back(stk.top());
        stk.pop();
      }
    }

    return res;
  }
};

int main() {
  vector<tuple<string, string>> tests {
    {"zza", "azz"},
    {"bac", "abc"},
    {"bdda", "addb"},
  };

  for (auto &[s, ans] : tests) {
    assert(Solution().robotWithString(s) == ans);
  }
}